### DISCUSSION TITLE
Name filters for duplicate devices

### DISCUSSION THREAD

**Discussion by bjorn-jaes** [Open-Agriculture/AgIsoStack-plus-plus#513] [Q&A] [answered]
Hello there!

Just wanted to ask if the `create_partnered_control_function()` could be used as follows:

``` C++
const std::vector<NAMEFilter> filters{/* some filters */};
auto PartnerCF1 = isobus::CANNetworkManager::CANNetwork.create_partnered_control_function(0, filters);
auto PartnerCF2 = isobus::CANNetworkManager::CANNetwork.create_partnered_control_function(0, filters);
```
to create partner CFs for devices that have the same filters without using their identity number as a filter?

I suspect that this is not the case, and that both the partner CFs would point to the same device, or something like that. What I probably need to do instead is use `get_control_functions()` to get all the control functions. Find the control functions that match the filters, then use their identity numbers as an additional filter for each partner. (Slight mention to [here](https://github.com/Open-Agriculture/AgIsoStack-plus-plus/discussions/483#:~:text=Yeah%2C%20there%27s%20a,seeder%2C%20sprayers%2C%20etc.))

**Comment by ad3154** [comment, marked as answer]
The goal is for what you stated above to work so that you can make multiple partners with the same filters. If that does not work, it would be a bug. Partners created with the same filters were not intended to point to the same control function.

**Reply by bjorn-jaes** [reply]
Thank you for the clarification!

I was uncertain after reading the documentation and some of the source code for the network manager. If you dont mind answering would it be possible to use a for/while loop to partner with all available devices of a given filter vector? 

I have some ideas:
1. Checking the name or address of the partner in a while loop, but from what I understand I would need to wait for an update first before checking those values.
2. Find a count using  `get_control_functions()` and then do a for loop.

Thank you for a much needed library!

**Reply by ad3154** [reply]
> If you dont mind answering would it be possible to use a for/while loop to partner with all available devices of a given filter vector

Sorry, didn't see this right away.

You can do both of those things, but you'd want to at least wait until your internal control function is valid (at most 1.5s after the bus is in good condition) to ensure that all control functions are registered. So, yes, but the be careful on the timing.
